### ISSUE TITLE
ImageMagick issues on PCs

### ISSUE THREAD

**Issue by hawc2** [YaleDHLab/pix-plot#41]
When running Pix-Plot on a PC, I'm getting this error: 

'identify' is not recognized as an internal or external command,
operable program or batch file.
'identify' is not recognized as an internal or external command,
operable program or batch file.
'identify' is not recognized as an internal or external command,
operable program or batch file.
Traceback (most recent call last):
  File "utils/process_images.py", line 438, in <module>
    clusters=args.clusters, validate_files=args.validate_files)
  File "utils/process_images.py", line 62, in __init__
    self.validate_inputs(validate_files)
  File "utils/process_images.py", line 98, in validate_inputs
    print(message)
  File "C:\Python35\lib\encodings\cp437.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_map)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\u0301' in position 1397: character maps to <undefined>

**Comment by hawc2** [comment]
so Doug suggested replacing the imagemagick commands with magick as a prefix (e.g. magick identify ...)

That brings another error. I'm trying to upload the image of it

![error](https://user-images.githubusercontent.com/15621648/40332348-ce7c362a-5d21-11e8-979f-89c29e680a70.jpg)

**Comment by duhaime** [comment]
@hawc2 Thanks for your patience with this Alex! Today I finally had a chance to sit down with a Windows OS and take a look at this. I've just merged a pull request that lets me run the data processing code on Windows 7.

Can I please ask you to try a fresh clone (or `git pull origin master -X theirs`), pip install the dependencies (ideally inside a conda environment), then retry processing your data? If you could try that and let us know how it goes I'd be grateful!

**Comment by duhaime** [comment]
@hawc2 Have you had a chance to try these changes out? If so, I'd be grateful if you could let us know how they went!

**Comment by hawc2** [comment]
Hey doug, it worked when I tested it with a small image set. The only error
we got was an ERRNO9 Bad File Descriptor, but it still worked.

However, I've been trying to test it on our set of 130000 images with the
PC, and one strange issue I started having was after starting the
processing, it seemed to be hanging without giving an error message or
processing the images. I just figured out it needed me to press return
again to keep going, and then again once it started to run the images. Not
sure if this is a new command you inserted or what. I'll let you know if
the full set of images properly process!

**Comment by duhaime** [comment]
@hawc2 That's quite interesting. I didn't process that many images when I made the Windows changes, but I'm not sure what might be causing that behavior. Do you happen to have a screenshot or log capture of where you needed to press Enter to continue the processing?

If you also happen to have a text log or screenshot of where the Bad File Descriptor warning sprang, I'd be grateful if you could post it and/or send the image that caused that warning!

**Comment by hawc2** [comment]
I'll be waiting like a day for the 130,000 images to process, but once
that's done I'll get some screenshots to share of the other errors

**Comment by duhaime** [comment]
@hawc2 Awesome, that sounds great! Thanks again for following up on this...

**Comment by hawc2** [comment]
So we ran into a few errors during this round. The first attached
screenshot, called Bad File Descriptor, will show you that error we
received. You can also see in this screenshot the final error with yet
another encoding error that stopped the processing. I'm pretty sure you may
have fixed the encoding errors for accent aigus, but not accent graves. The
image that caused the error included an accent grave. Anyway, we're
changing the image names and running it again.

We also received an error message relating to Tesnorflow's CPU usage you
can see in the second attached screenshot. That didn't stop the processing,
but seems worth noting

Finally, regarding the error where the program hangs at the stop, until I
press enter: well, I tried it again, and this time when I pressed enter it
didn't get it going. I don't know if it needs more time to process or what,
but it's weird. IIt hangs after it says "Validating input files"

**Comment by hawc2** [comment]
ok uploaded the images. and yeah this is the most recent commit, pulled a couple days ago. and yeah we did check for processing usage, and we hadn't hit peak ram. this computer should be able to handle it

**Comment by pleonard212** [comment]
I'm seeing three things to try and check:

1) spaces in directory names `Book Covers` and `Full set`
2) unicode in file names: `Agnès_Laurent.jpg`, which is maybe OK but is triggering an escaped character string that starts with a Windows directory separator (`Agne\u0300s`)
3) mixed Windows/UNIX directory separators `Book Covers/Full set\Agnes`...

Some of these we can catch with code but I'm wondering -- for now, could you rename Agnès to Agnes and take out the spaces in those dirs to see what happens?

**Comment by hawc2** [comment]
Yeah the spaces I had tested and weren't a problem. I changed all the file
names and am going to run that now. The accent grave must have been the
problem with the unicode. the mixed separators I had missed so that might
have been an issue. Will let you know what happens, but it will  take a
while to process...

**Comment by duhaime** [comment]
Yes, it seems the non-unicode filename is causing Windows to choke. One can run ImageMagick commands on files with non-unicode characters:

![woohoo-paint](https://user-images.githubusercontent.com/4801116/40791256-168b7412-64c5-11e8-9b24-08c44fd82dcb.png)

But when trying to process images with non-unicode characters on Windows pixplot flails:

![hello](https://user-images.githubusercontent.com/4801116/40791368-54316696-64c5-11e8-94ac-5269a37068ea.png)

That said, all other files in the directory (even those with spaces) were processed just fine...

**Comment by duhaime** [comment]
It looks like the non-ascii filename problem may be resolved in more recent Tensorflow versions: https://github.com/tensorflow/tensorflow/pull/11562

**Comment by duhaime** [comment]
@hawc2 Would you be able to try pulling down some recent changes to see if they might help clean up the data processing on Windows? To pull down the branch, you can run:

```
git pull origin tensorflow1.8
```

If you pull that down and run it on your images, I'd be grateful if you could let us know how it goes...

**Comment by duhaime** [comment]
The bad file descriptor message seems to be caused by using only a single directory delimiter instead of two directory delimiters. That is to say, while `"D:\Users\Tony\bingo\*.jpg"` throws a bad file descriptor, `"D:\\Users\\Tony\\bingo\\*.jpg"` does not.

So instead of:

```bash
python utils\process_images.py "C:\Users\ded34\Desktop\pix-plot\hello world\*"
```
one should do:
```bash
python utils\\process_images.py "C:\\Users\\ded34\\Desktop\\pix-plot\\hello world\\*"
```

**Comment by hawc2** [comment]
Hey Doug, sorry for the delay - we have gotten it working on the PC. The
issue with the program hanging before starting processing we figured out
too. That was it taking a very very long time to process the verification
step. It was strange how long it would hang, sometimes for more than a
work-day, before running. Since we already had the images verified, we
skipped that step.

By modifying the parameters, we were able to get slightly better signal,
but still not a wide distribution and separated clustering. Our next step
is to get a set of book covers (of any genre) that are better resolution to
see what happens. I'll let you know how that goes!

Thanks for all your help!

Best,
Alex
